blobfuse/OAuthToken.cpp

```diff
@@ -44,60 +51,137 @@ void from_json(const json &j, OAuthToken &t) {
-    try {
-        if (j.contains("expires_on"))
+    if (!expin_failed) // if there is an expires_in just use it and dont worry parsing expires_on field
+    {
+        time_t current_time;
+            
+        struct tm *temptm;
+            
+        time(&current_time);
+            
+        temptm = gmtime( &current_time );
+
+        current_time = mktime(temptm);
+
+        t.expires_on = current_time + t.expires_in;    
+        
+        syslog(LOG_WARNING, "After adding specified expires_in token expiry time in utc %s\n", ctime(&t.expires_on));
+    }
+    else if (j.contains("expires_on"))
+    {
+        try 
         {
             auto val = j.at("expires_on");
             if(val.is_number())
             {
+                // if the expires_on it is localtime. Do not convert it.
+                syslog(LOG_INFO, "expires_on date is a number so using val_get_to to convert it");
                 val.get_to(t.expires_on);
             }
-            else {
+            else 
+            {
                 std::string expires_on = val.get<std::string>();
-                t.expires_on = std::stoi(expires_on);
+                
+                if (is_dt_number(expires_on)) // check with the custom method as the above does not catch everything
+                {
+                    t.expires_on = std::stoi(expires_on);
+                }    
+                else 
+                { // now the date is a string in either the UTC or localtime format so just parse it
+                    fprintf(stdout, "expires_on date is string");
+                    
+                    
+                    // try UTC first then use else to implement in local time in AM or PM
+                    int utcIndex = expires_on.find("+0000");
+                    if ( utcIndex > 19)
+                    { 
+                    // remove tiemzone
+                        expires_on = expires_on.substr(0, utcIndex);
+                        // remove the trailing space if any
+                        expires_on.erase(std::find_if(expires_on.rbegin(), expires_on.rend(), std::bind1st(std::not_equal_to<char>(), ' ')).base(), expires_on.end());
+                        // Now remove the milliseconds because strptime is going crazy with those milliseconds anyway we dont use them so.
+                        int millisecondsIndex  =  expires_on.find(".");
+                        if (millisecondsIndex > 18)
+                        {
+                            expires_on = expires_on.substr(0, millisecondsIndex);
+                        }
+                        struct tm timeStruct;
+                    // Ref: for formats: https://www.tutorialcup.com/cplusplus/date-time.htm
+                        if ((strptime(expires_on.c_str(), "%Y-%m-%d %H:%M:%S", &timeStruct) != NULL)
+                            ||
+                            (strptime(expires_on.c_str(), "%Y-%b-%d %H:%M:%S", &timeStruct) != NULL))                  

```

Comment:
NIT - Could the || be moved to the same line as the first line of the if statement?

Reply:
easier to read this way